### PR TITLE
refactor runner terminal summary and add more durations coverage

### PR DESCRIPTION
**Pull request by symonk**
Hi all, while investigating #8042 (Awaiting response from the author of that issue before I invest more time investigating as I am unable to recreate) I did a bit of tidy up in the `terminal_summary` hook of the `runner.py` plugin. 

- Improved (in my opinion) local scope var names to be a little more readable.
- replace the sorting with attrgetter vs the anonymous lambda function.
- Move to the preferred getOption(attr) calls via the config obj, rather than the option.attr access
- Pluralize the output to handle the edge case of `1 duration` vs `x durations`.
- Added some extra test coverage around `--durations`.
- slightly more readable slower than output message, including formatting to 3f (in-line with --duration-min default `0.005`.

I plan to fix the issue of 8042 if it is recreatable on a windows environment, unable to recreate so that will be a separate PR.

Thanks.

### CODE REVIEW

**Comment by symonk** [comment]
Swapping from lambda to attrgetter is maybe over zealous for no gain really and an extra import, thoughts on that? Leaning towards putting the lambda back here

File: testing/acceptance_test.py [hunk 1]
```diff
@@ -1288,3 +1290,62 @@ def test_no_brokenpipeerror_message(pytester: Pytester) -> None:
...
+        @pytest.mark.parametrize('_', range(6))
+        def test_this(_):
+            time.sleep(0.1)
+        """
+    )
+    result = pytester.runpytest("--durations=0")
+    assert result.ret == ExitCode.OK
+    # Check 6x setup & 6x teardown are hidden, but 6x call are shown
+    insertable_line = "*test_durations_zero_lists_all.py::test_this*{}*"
+    lines = [insertable_line.format(x) for x in range(6)]
```
Comment (nicoddemus) [inline testing/acceptance_test.py:1347]: In my experience tests that depend on a tight sleep like this one tend to be flaky; sometimes an overloaded CI machine will take a bit longer to run a setup function (CI machines tend to have shared CPUs) and then the test will fail.

Don't have a specific suggestion to improve this right now, I suggest we keep this in mind and wait and see.

**Review by nicoddemus — APPROVED** [review]
Thanks!

About the lambda vs operator.getter, I tend to use the latter but I think it is mainly a personal choice, so go with whatever you prefer.

File: testing/acceptance_test.py (hunk 1 above)
Comment (symonk) [reply testing/acceptance_test.py:1347]: Yeah I was kind of confused in how best to test it (reliably), I can increase the sleep in the test(s) themselves but at the price of 'dead time' in the test run, that said I wouldn't want to introduce something very flaky as its a massive time headache later, open to suggestions

Comment (nicoddemus) [reply testing/acceptance_test.py:1347]: One way would be to make unit-tests instead of integration tests; that would require to refactor `pytest_terminal_summary` to extract the relevant logic about building up the list report with pre-defined durations which don't depend on sleep, and then test that.

Comment (symonk) [reply testing/acceptance_test.py:1347]: will give it a go, any advice on how to proceed? refactoring a hook seems like the knock on effects could be collossal

Comment (nicoddemus) [reply testing/acceptance_test.py:1347]: > will give it a go, any advice on how to proceed? refactoring a hook seems like the knock on effects could be collossal

I meant to just refactor the `pytest_terminal_hook` _implementation_ in `terminal.py`, not to change the `pytest_terminal_hook` spec itself, so the changes should be local and not affect anything else other than `terminal.py`.

File: src/_pytest/runner.py [hunk 2]
```diff
@@ -64,35 +65,43 @@ def pytest_addoption(parser: Parser) -> None:
...
+    durations_list = []
+    for report_list in reporter.stats.values():
+        for report in report_list:
+            if hasattr(report, "duration"):
+                durations_list.append(report)
+    if not durations_list:
         return
-    dlist.sort(key=lambda x: x.duration, reverse=True)  # type: ignore[no-any-return]
+
+    durations_list.sort(key=attrgetter("duration"), reverse=True)
```
Comment (bluetech) [inline src/_pytest/runner.py:83]: I prefer the lambda only because `attrgetter` gets the field name as a string which means it cannot be typed.

File: src/_pytest/runner.py [hunk 3]
```diff
@@ -64,35 +65,43 @@ def pytest_addoption(parser: Parser) -> None:
...
-            tr.write_line(
-                "(%s durations < %gs hidden.  Use -vv to show these durations.)"
-                % (len(dlist) - i, durations_min)
+        reporter.write_sep("=", "slowest %s durations" % durations)
+        durations_list = durations_list[:durations]
+
+    for index, test_report in enumerate(durations_list):
+        if verbose < 2 and test_report.duration < durations_min:
+            total = len(durations_list) - index
+            pronoun, dur = (
```
Comment (bluetech) [inline src/_pytest/runner.py:94]: I suggest `dur` -> `subject`

**Review by bluetech — APPROVED** [review]
Nice improvement @symonk. 

I agree with @nicoddemus about the `sleep`s, they will likely be flaky, in addition to slowing down the tests slightly, but if you don't get around to finding a better solution I think we can still merge, but remove/adjust the tests if they prove to be flaky.

File: src/_pytest/runner.py (hunk 2 above)
Comment (bluetech) [reply src/_pytest/runner.py:83]: (I mean type *checked*)

**Comment by nicoddemus** [comment]
Hi folks, anything else needed here (other than fixing the conflicts)?

**Comment by symonk** [comment]
@nicoddemus / @bluetech thanks for the reviews here, ultimately I am not happy merging something tested like this into `main`, i think it will be an injustice later and create some development headache with flakyness, I plan to revisit this PR and rewrite the `pytest_terminal_summary` into something more unit-testable and write some more robust coverage here, its my current short term goal in pytest anyway.  Thanks for the reviews thus far, hopefully it won't take too long and sorry for the delay

**Comment by obestwalter** [comment]
@symonk would you like to finish this up? Or maybe somebody else? Otherwise, I think we might rather want to close this for now.

**Comment by bluetech** [comment]
Closing as stale per comment above.
